### PR TITLE
more modernization of C-Fortran interface

### PR DESCRIPTION
**Pull request by jbathegit**
Part of #79

Modernize interface for cadn30, maxbyt, numtbd, igetntbi, elemdx, and parts of restd and stseq.

Part of #252

Combines cmpia, nummtb, and stseq into a single C file.

### CODE REVIEW

File: src/bufr_c2f_interface.F90 [hunk 1]
```diff
@@ -280,7 +285,7 @@ end subroutine mtinfo_c
     !> Wraps status() subroutine.
     !>
     !> @param file_unit - Fortran logical unit number of file.
-    !> @param lun - file ID.
+    !> @param lun - File ID.
     !> @param il - file status.
     !> @param im - message status.
```
Comment (AlexanderRichert-NOAA) [inline src/bufr_c2f_interface.F90]: Would it make sense to capitalize 'file status' and 'message status' while we're at it?

File: src/bufr_interface.h [hunk 2]
```diff
@@ -160,7 +160,7 @@ extern "C" {
  * Wraps status() subroutine.
  *
  * @param file_unit - Fortran logical unit number of file.
- * @param lun - file ID.
+ * @param lun - File ID.
  * @param il - file status.
  * @param im - message status.
```
Comment (AlexanderRichert-NOAA) [inline src/bufr_interface.h]: " "

File: src/bufr_interface.h [hunk 3]
```diff
@@ -207,39 +207,39 @@ extern "C" {
...
 
 /**
  * Get information about a descriptor.
  *
  * Wraps nemtab() subroutine.
  *
- * @param bufr_unit - the bufr file pointer.
+ * @param bufr_unit - File ID.
  * @param mnemonic - mnemonic.
  * @param descriptor - the binary descriptor for the mnemonic.
```
Comment (AlexanderRichert-NOAA) [inline src/bufr_interface.h]: " "

File: src/bufr_interface.h [hunk 4]
```diff
@@ -333,7 +333,7 @@ extern "C" {
 /**
  * Get the bufr node idx for the start node of the subset.
  *
- * @param lun - pointer for the file stream.
+ * @param lun - File ID.
```
Comment (AlexanderRichert-NOAA) [inline src/bufr_interface.h:336]: Do 'bufr_unit' and 'lun' basically mean the same thing?

**Comment by AlexanderRichert-NOAA** [comment]
Looking at the diffs, there are some places where your name is 'J Ator' or 'J.Ator', I'll leave it to you if you want to search/replace those.

File: src/bufrlib.h.in [hunk 5]
```diff
@@ -49,6 +47,63 @@ int icvidx(int ii, int jj, int numjj);
  * given time for a Table B descriptor.*/
 #define MXNAF @MXNAF@
 
+/**
+ * Convert an FXY value from its bit-wise (integer) representation to its
```
Comment (edwardhartnett) [inline src/bufrlib.h.in]: WE are calling this a "WMO bit-wise representation of the FXY value" and adding a reference to the ifxy() function, which explains it fully. See adn30.f for example.

File: src/restd.c [hunk 6]
```diff
@@ -22,33 +22,36 @@
...
+ * to tddesc.
  *
  * @author J. Ator @date 2004-08-18
 */
 
 void
 restd(f77int *lun, f77int *tddesc, f77int *nctddesc, f77int *ctddesc)
 {
+    static const int ADN_STR_LEN = 7;
+    static const int NEMO_STR_LEN = 9;
```
Comment (edwardhartnett) [inline src/restd.c]: I'm very glad these are now deffined, because we should not have naked constants in our code.

However, in C these would typically be defined with the preprocessor. The reason is that many such constants will end up migrating up to a .h file, when you find that you are using the same constant in multiple places.

This has the minor advantage of not putting them on the stack at all, since the pre-processor is run before compilation.

So instead of defining static ints, use:

```
#define ADN_STR_LEN 7
#define NEMO_STR_LEN 9

```

Did you include the NULL char in the string len? If so, the convention is not to. In other words:

```
#define ADN_STR_LEN 6
char adn_str[ADN_STR_LEN + 1];

```
In other words, make the STR_LEN equal to the length of the string without the NULL, and make sure you add the extra +1 whenever you allocate storage. It might seem funny, but this convention makes everything work well with the C string library, which uses the same style.

File: src/restd.c [hunk 7]
```diff
@@ -61,7 +64,7 @@ restd(f77int *lun, f77int *tddesc, f77int *nctddesc, f77int *ctddesc)
 /*
 **          desc is a local descriptor.
 */
-            numtbd( lun, &desc, nemo, &tab, &ictbd, 9, 1 );
+            numtbd_f( *lun, desc, nemo, NEMO_STR_LEN, &tab, &ictbd );
```
Comment (edwardhartnett) [inline src/restd.c]: Suggest you remove the spaces after and before parens. Not usual in C code.

File: src/stseq.c [hunk 8]
```diff
@@ -8,6 +8,82 @@
...
+ * @author J Ator @date 2009-03-23
+*/
+void
+nummtb(int *idn, char *tab, int *ipt)
+{
+        int *pifxyn, *pbs,  nmt;
+
+        char adn[FXY_STR_LEN], errstr[129];
+
+        if ( *idn >= ifxy( "300000", 6 ) ) {
```
Comment (edwardhartnett) [inline src/stseq.c]: What's happening here? What is "3000000"? Is that an FXY that is too large? A comment would be helpful and the string "3000000" should be defined as a pre-processor constant instead of being a naked constant in the code.

File: src/stseq.c [hunk 9]
```diff
@@ -19,30 +95,32 @@
  * are themselves Table D descriptors are automatically resolved via
  * a recursive call to this same subroutine.
  *
- * @param[in] lun - File ID.
- * @param[in,out] irepct - Replication sequence counter for the current
+ * @param lun - File ID.
+ * @param irepct - Replication sequence counter for the current
```
Comment (edwardhartnett) [inline src/stseq.c:96]: I usually phrase it like this:

`@param irepct - Pointer to the replication sequence counter for the current
`

File: src/bufr_interface.h (hunk 4 above)
Comment (jbathegit) [reply src/bufr_interface.h:336]: bufr_unit is typically the Fortran logical unit number on the system that's associated with a file, whereas lun is an internal (to the library) index used to keep track of all such logical unit numbers that have already been introduced to the library via calls to openbf.  So they are different entities, and we agreed in a previous discussion to refer to lun as "File ID" going forward, instead of some of the other descriptions (e.g. I/O stream index, etc.) that we've used in the past.

We're trying to update all of these as we go forward, but as you can see there are still some left to go, and I see now that the bufr_file variable name for nemtab and nemtbb in this file should really be lun, so I'll get that cleaned up.

File: src/bufr_c2f_interface.F90 (hunk 1 above)
Comment (jbathegit) [reply src/bufr_c2f_interface.F90]: Probably, if no other reason than for consistency with the other argument descriptions.  I'll do that.

File: src/bufrlib.h.in (hunk 5 above)
Comment (jbathegit) [reply src/bufrlib.h.in]: I know we discussed changing this description, but from what I recall we were going to think some more about what a suitable replacement would be, and I didn't think we'd decided on anything yet.

That said, I'm OK with this new description, so I'll go ahead and fix that.

File: src/restd.c (hunk 6 above)
Comment (jbathegit) [reply src/restd.c]: I did include NULL in these string lengths, and I could redefine everything with the +1 to make it clearer, but FYI that the allocated size also needs to be passed as an argument to the _f routines that are called by this routine.  It's needed for use within copy_f_c_str which is called inside of those _f wrapper routines, for copying between Fortran and C strings to make sure we don't overflow the allocated C space.  So this means I'd also need to include the +1 in those arguments, and not just when initially allocating the storage.

File: src/stseq.c (hunk 8 above)
Comment (jbathegit) [reply src/stseq.c]: No, this is a test to determine whether it's a Table D descriptor (whose FXY values always have a "3" as their first digit) or a Table B descriptor (whose FXY values always have a "0" as their first digit).  IMHO, the following lines `tab = 'D'` and `tab = 'B'` already clarify this, but if you want I can add a few more comments.  Either way, I don't think this is really a "naked constant" in the sense that you're thinking.

File: src/restd.c (hunk 7 above)
Comment (jbathegit) [reply src/restd.c]: OK will do.

**Comment by jbathegit** [comment]
Thanks for all the feedback!  I think I've covered everything with my latest commit, but please let me know if you have any further questions or comments.

**Review by AlexanderRichert-NOAA — APPROVED** [review]
Tests pass on my personal machine with gcc
